src/connectionController.ts

```diff
@@ -116,142 +122,188 @@ export default class ConnectionController {
     this._telemetryService = telemetryService;
   }
 
-  async _migratePreviouslySavedConnection(
-    savedConnectionInfo: StoreConnectionInfo
-  ): Promise<StoreConnectionInfoWithConnectionOptions> {
-    if (!savedConnectionInfo.connectionModel) {
-      throw new Error(
-        'The connectionModel object is missing in saved connection info.'
+  async loadSavedConnections(): Promise<void> {
+    const globalAndWorkspaceConnections = {
+      ...this._storageController.get(
+        StorageVariables.GLOBAL_SAVED_CONNECTIONS,
+        StorageLocation.GLOBAL
+      ),
+      ...this._storageController.get(
+        StorageVariables.WORKSPACE_SAVED_CONNECTIONS,
+        StorageLocation.WORKSPACE
+      ),
+    };
+
+    let connectionsDidChange = false;
+    const connectionsWithSecretsInKeytar: Array<{
+      connectionId: string;
+      connectionName: string;
+    }> = [];
+    const totalConnectionEntries = Object.entries(
+      globalAndWorkspaceConnections
+    );
+    for (const [connectionId, connectionInfo] of totalConnectionEntries) {
+      const connectionInfoWithSecret = await this._getConnectionInfoWithSecrets(
+        connectionInfo
+      );
+
+      if (connectionInfoWithSecret) {
+        connectionsDidChange = true;
+        this._connections[connectionId] = connectionInfoWithSecret;
+
+        if (
+          connectionInfoWithSecret.secretStorageLocation ===
+          SecretStorageLocation.Keytar
+        ) {
+          connectionsWithSecretsInKeytar.push({
+            connectionId,
+            connectionName: connectionInfo.name,
+          });
+        }
+      }
+    }
+
+    if (connectionsDidChange) {
+      this.eventEmitter.emit(DataServiceEventTypes.CONNECTIONS_DID_CHANGE);
+    }
+
+    if (connectionsWithSecretsInKeytar.length) {
+      log.error(
+        'Could not migrate secrets for a few connections',
+        connectionsWithSecretsInKeytar
+      );
+      this._telemetryService.track(
+        TelemetryEventTypes.KEYTAR_SECRETS_MIGRATION_FAILED,
+        {
+          totalConnections: totalConnectionEntries.length,
+          connectionsWithSecretsInKeytar: connectionsWithSecretsInKeytar.length,
+        }
+      );
+      void vscode.window.showInformationMessage(
+        [
+          'Could not migrate secrets for a few connections. Please review the following connections:',
+          connectionsWithSecretsInKeytar
+            .map(({ connectionName }) => connectionName)
+            .join(', '),
+        ].join('\n')
       );
     }
+  }
+
+  async _getConnectionInfoWithSecrets(
+    connectionInfo: StoreConnectionInfo
+  ): Promise<MigratedStoreConnectionInfoWithConnectionOptions | undefined> {
+    try {
+      if (connectionInfo.connectionModel) {
+        return this._migratePreviouslySavedConnection(
```

Comment:
Maybe we should rename this given that there are multiple different kinds of migrations now?

Reply:
Wanted to suggest the same. It can be something like `_migrateConnectionsWithConnectionModel` and `_migrateConnectionsWithKeytarSecrets` or similar to indicate that in both cases we update connections based on different attributes.